core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -528,7 +528,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       }
 
     // Optionally scale number of executors dynamically based on workload. Exposed for testing.
-    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false)
+    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false) &&
+      (_conf.getInt("spark.executor.instances", 0) == 0)
```

Comment:
since this is now duplicated in a few places, maybe it makes sense to make a `Utils.isDynamicAllocationEnabled` method and move the comment there instead.